mne/io/nirx/nirx.py

```diff
@@ -0,0 +1,266 @@
+        # The source location is stored in the second 3 entries of loc.
+        # The detector location is stored in the third 3 entries of loc.
+        # NIRx NIRSite uses MNI coordinates.
+        for ch_idx2 in range(requested_channels.shape[0]):
+            # Find source and store location
+            src = int(requested_channels[ch_idx2, 0]) - 1
+            info['chs'][ch_idx2 * 2]['loc'][3:6] = src_locs[src, :]
+            info['chs'][ch_idx2 * 2 + 1]['loc'][3:6] = src_locs[src, :]
+            # Find detector and store location
+            det = int(requested_channels[ch_idx2, 1]) - 1
+            info['chs'][ch_idx2 * 2]['loc'][6:9] = det_locs[det, :]
+            info['chs'][ch_idx2 * 2 + 1]['loc'][6:9] = det_locs[det, :]
+            # Store channel location
+            # Channel locations for short channels are bodged,
+            # for short channels use the source location and add small offset
+            if (has_short > 0) & (len(np.where(short_det == det + 1)[0]) > 0):
+                info['chs'][ch_idx2 * 2]['loc'][:3] = src_locs[src, :]
+                info['chs'][ch_idx2 * 2 + 1]['loc'][:3] = src_locs[src, :]
+                info['chs'][ch_idx2 * 2]['loc'][0] += 0.8
+                info['chs'][ch_idx2 * 2 + 1]['loc'][0] += 0.8
+            else:
+                info['chs'][ch_idx2 * 2]['loc'][:3] = ch_locs[ch_idx2, :]
+                info['chs'][ch_idx2 * 2 + 1]['loc'][:3] = ch_locs[ch_idx2, :]
+
+        raw_extras = {"sd_index": req_ind, 'files': files}
+
+        super(RawNIRX, self).__init__(
+            info, preload, filenames=[fname], last_samps=[last_sample],
+            raw_extras=[raw_extras], verbose=verbose)
+
+        # Read triggers from event file
+        t = [re.findall(r'(\d+)', line) for line in open(files['evt'])]
+        onset = np.zeros(len(t), float)
+        duration = np.zeros(len(t), float)
+        description = [''] * len(t)
+        for t_idx in range(len(t)):
+            binary_value = ''.join(t[t_idx][1:])[::-1]
+            trigger_frame = float(t[t_idx][0])
+            onset[t_idx] = (trigger_frame) * (1.0 / samplingrate)
+            duration[t_idx] = 1.0  # No duration info stored in files
+            description[t_idx] = int(binary_value, 2) * 1.
+        annot = Annotations(onset, duration, description)
+        self.set_annotations(annot)
+
+    def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
+        """Read a segment of data from a file.
+
+        The NIRX machine records raw data as two different wavelengths.
+        The returned data interleaves the wavelengths.
+        """
+        file_wl1 = self._raw_extras[fi]['files']['wl1']
+        file_wl2 = self._raw_extras[fi]['files']['wl2']
+
+        sdindex = self._raw_extras[fi]['sd_index']
+
+        wl1 = np.loadtxt(file_wl1, delimiter=' ')
+        wl2 = np.loadtxt(file_wl2, delimiter=' ')
+
+        wl1 = wl1[start:stop, sdindex].T
+        wl2 = wl2[start:stop, sdindex].T
```

Comment:
you read here the whole file all the time. You should read only the lines you need. loadtxt can do this AFAIK

Reply:
It does have `skiprows` and `usecols` but nothing for the number of rows to read.

You are probably thinking of `pandas.read_csv`, which has both `skiprows=None` and `nrows=None` :)

Staying with `numpy` a fast way to do this would be on init, read through the file to get the offsets of the lines. Then when you want to read a chunk, `fid.read()` from one part to another, and pass this chunk to `numpy.fromtxt` directly. Actually it might not be so bad...

Reply:
Also the NumPy code is mostly pure Python anyway, so I just tried to write something that will be quick in Python. (It seemed to be at least as fast as `pandas` in basic testing, but we can always optimize later.)